### PR TITLE
Added support for oracle databases

### PR DESCRIPTION
**Pull request by 15168326318**
是基于develop 0.4.1-SNAPSHOT开发的

### CODE REVIEW

**Comment by xingfudeshi** [comment]
Nice work.please sign CLA first，so we can accept yor PR.

**Comment by xingfudeshi** [comment]
I left some comments

**Comment by sharajava** [comment]
Great! 👍

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    keywordChecker=new ORACLEKeywordChecker();
+                }
+            }
+        }
+        return keywordChecker;
+    }
+    /**
+     * MySQL keyword
+     */
+    private enum MySQLKeyword {
```
Comment (sharajava) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java]: ORACLE?

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/ORACLEDeleteRecognizer.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import com.alibaba.fescar.rm.datasource.sql.SQLDeleteRecognizer;
+import com.alibaba.fescar.rm.datasource.sql.SQLType;
+import com.alibaba.fescar.rm.datasource.sql.druid.BaseRecognizer;
+
+import java.util.ArrayList;
+
+/**
+ * The type My sql delete recognizer.
+ */
+public class ORACLEDeleteRecognizer extends BaseRecognizer implements SQLDeleteRecognizer {
```
Comment (slievrly) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/ORACLEDeleteRecognizer.java]: add author

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (15168326318) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java]: 已经改为oracle了

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        MySQLKeyword(String name) {
+            this.name = name;
+        }
+    }
+
+
+    @Override
+    public boolean check(String fieldOrTableName) {
+        try {
```
[the remark this answers is not in the document]
Comment (15168326318) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/oracle/keyword/ORACLEKeywordChecker.java]: 已经按0.4.1快走版本修改了。

**Comment by 15168326318** [comment]
冲突已经解决，并且注释，不规范的命名都该好了。

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/exec/BaseTransactionalExecutor.java [hunk 4]
```diff
@@ -20,15 +20,13 @@
...
 import com.alibaba.fescar.rm.datasource.ConnectionProxy;
 import com.alibaba.fescar.rm.datasource.StatementProxy;
 import com.alibaba.fescar.rm.datasource.sql.SQLRecognizer;
 import com.alibaba.fescar.rm.datasource.sql.SQLType;
-import com.alibaba.fescar.rm.datasource.sql.struct.Field;
-import com.alibaba.fescar.rm.datasource.sql.struct.TableMeta;
-import com.alibaba.fescar.rm.datasource.sql.struct.TableMetaCache;
-import com.alibaba.fescar.rm.datasource.sql.struct.TableRecords;
+import com.alibaba.fescar.rm.datasource.sql.struct.*;
 import com.alibaba.fescar.rm.datasource.undo.SQLUndoLog;
```
Comment (lovepoem) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/exec/BaseTransactionalExecutor.java]: .* is an incorrect  type

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/OracleSelectForUpdateRecognizer.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ *  limitations under the License.
+ */
+
+package com.alibaba.fescar.rm.datasource.sql.druid.oracle;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOpExpr;
+import com.alibaba.druid.sql.ast.expr.SQLVariantRefExpr;
+import com.alibaba.druid.sql.ast.statement.*;
```
Comment (lovepoem) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/OracleSelectForUpdateRecognizer.java]: .* is an incorrect type

Comment (15168326318) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/OracleSelectForUpdateRecognizer.java]: idea自己会引入这样的。

**Comment by 15168326318** [comment]
.* is an incorrect type，这个已经改好了。

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/OracleUpdateRecognizer.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.alibaba.fescar.rm.datasource.sql.druid.oracle;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.SQLStatement;
+import com.alibaba.druid.sql.ast.expr.*;
```
Comment (xingfudeshi) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/sql/druid/oracle/OracleUpdateRecognizer.java]: please import explicit class .should not use import *.

**Comment by 15168326318** [comment]
@xingfudeshi @lovepoem ,你们提的changes requested我已经改好了。

**Comment by xingfudeshi** [comment]
@15168326318 one more thing,please update the PR title to English.it will be the commit log.

**Comment by 15168326318** [comment]
@xingfudeshi 标题已经改为英文了。

**Comment by xingfudeshi** [comment]
@15168326318 OK

**Comment by slievrly** [comment]
Expected to be merge at 0.8

**Comment by johnlin08** [comment]
what a nice job!!!But OracleKeywordChecker doesn't support oracle 'merge into'.  And does it support batch insert or update?

**Comment by johnlin08** [comment]
@15168326318

**Comment by johnlin08** [comment]
@15168326318  What a nice job!!!But OracleKeywordChecker doesn't support oracle 'merge into'. And does it support batch insert or update?

**Comment by GFriedrich** [comment]
@xingfudeshi What happened to the Oracle feature branch? I can't find it or is it hidden somewhere?

**Comment by lovepoem** [comment]
>  @xingfudeshi What happened to the Oracle feature branch? I can't find it or is it hidden somewhere?

The seata: feature_oracle is deleted by mistake

@15168326318 Would you please submit a new pr to develop branch ?

**Pull request merged**